crates/motoko/src/lib/vm.rs

```diff
@@ -1284,28 +1258,41 @@ fn stack_cont(core: &mut Core, v: Value) -> Result<Step, Interruption> {
                 _ => Err(Interruption::TypeMismatch),
             },
             Opt => {
-                core.cont = Cont::Value(Value::Option(Box::new(v)));
+                core.cont = Cont::Value(Value::Option(Rc::new(v)));
                 Ok(Step {})
             }
             DoOpt => {
-                core.cont = Cont::Value(Value::Option(Box::new(v)));
+                core.cont = Cont::Value(Value::Option(Rc::new(v)));
                 Ok(Step {})
             }
             Bang => match v {
                 Value::Option(v) => {
-                    core.cont = Cont::Value(*v);
+                    core.cont = Cont::Value((*v).clone());
                     Ok(Step {})
                 }
                 Value::Null => bang_null(core),
                 _ => Err(Interruption::TypeMismatch),
             },
-            Call1(inst, e2) => match v {
-                Value::Function(cf) => exp_conts(core, FrameCont::Call2(cf, inst), e2),
-                Value::PrimFunction(pf) => exp_conts(core, FrameCont::Call2Prim(pf, inst), e2),
+            Call1(inst, e2) => {
+                let v = core.deref_value(Rc::new(v))?;
```

Comment:
Introducing a new `Rc::new` here is not "sharing" the original one that was surrounding with `v`.

So, it's not actually doing what we want, if the point is to maximize sharing in the stack frame structure.

Reply:
Perhaps this is related to the fairly large regression in performance that I think is attributable to this PR?